File: app/src/main/java/com/bunbeauty/ideal/myapplication/cleanArchitecture/ui/activities/logIn/VerifyPhoneActivity.java
package com.bunbeauty.ideal.myapplication.cleanArchitecture.ui.activities.logIn;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.android.ideal.myapplication.R;
import com.arellomobile.mvp.MvpAppCompatActivity;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.bunbeauty.ideal.myapplication.cleanArchitecture.business.logIn.VerifyPhoneInteractor;
import com.bunbeauty.ideal.myapplication.cleanArchitecture.callback.VerifyCallback;
import com.bunbeauty.ideal.myapplication.cleanArchitecture.di.AppModule;
import com.bunbeauty.ideal.myapplication.cleanArchitecture.di.DaggerAppComponent;
import com.bunbeauty.ideal.myapplication.cleanArchitecture.mvp.presenters.VerifyPhonePresenter;
import com.bunbeauty.ideal.myapplication.cleanArchitecture.mvp.views.VerifyPhoneView;
import com.bunbeauty.ideal.myapplication.helpApi.WorkWithViewApi;

import javax.inject.Inject;

public class VerifyPhoneActivity extends MvpAppCompatActivity implements View.OnClickListener, VerifyCallback, VerifyPhoneView {

    private static final String TAG = "DBInf";

    private Button verifyCodeBtn;
    private TextView resendCodeText;
    private EditText codeInput;
    private TextView changePhoneText;
    private TextView alertCodeText;
    private ProgressBar progressBar;

    @Inject
    VerifyPhoneInteractor verifyPhoneInteractor;

    @InjectPresenter
    VerifyPhonePresenter verifyPhonePresenter;

    @ProvidePresenter
    VerifyPhonePresenter provideVerifyPhonePresenter() {
        DaggerAppComponent.builder()
                .appModule(new AppModule(getApplication()))
                .build()
                .inject(this);
        return new VerifyPhonePresenter(verifyPhoneInteractor);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.verify_phone);

        initView();
        verifyPhonePresenter.sendCode(this);
    }

    private void initView() {
        Log.d(TAG, "initView RegistrationActivity: ");

        verifyCodeBtn = findViewById(R.id.verifyVerifyBtn);
        resendCodeText = findViewById(R.id.resendVerifyText);
        alertCodeText = findViewById(R.id.alertCodeVerifyText);
        codeInput = findViewById(R.id.codeVerifyInput);
        changePhoneText = findViewById(R.id.changePhoneVerifyText);

        progressBar = findViewById(R.id.progressBarVerifyCode);

        verifyCodeBtn.setOnClickListener(this);
        resendCodeText.setOnClickListener(this);
        changePhoneText.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        WorkWithViewApi.hideKeyboard(this);
        switch (v.getId()) {
            case R.id.verifyVerifyBtn:
                verifyPhonePresenter.verify(codeInput.getText().toString(), this);
                break;
            case R.id.resendVerifyText:
                verifyPhonePresenter.resendCode(this);
                break;
            case R.id.changePhoneVerifyText:
                goBackToAuthorization();
                break;
            default:
                break;
        }
    }

    @Override
    public void hideViewsOnScreen() {
        verifyCodeBtn.setVisibility(View.GONE);
        resendCodeText.setVisibility(View.GONE);
        codeInput.setVisibility(View.GONE);
        changePhoneText.setVisibility(View.GONE);
        alertCodeText.setVisibility(View.GONE);
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void showViewsOnScreen() {
        verifyCodeBtn.setVisibility(View.VISIBLE);
        resendCodeText.setVisibility(View.VISIBLE);
        codeInput.setVisibility(View.VISIBLE);
        changePhoneText.setVisibility(View.VISIBLE);
        alertCodeText.setVisibility(View.VISIBLE);
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void showResendCode() {
        Toast.makeText(this, "Код был отправлен", Toast.LENGTH_LONG).show();
    }

    @Override
    public void showWrongCode() {
        Toast.makeText(this, "Вы ввели неверный код", Toast.LENGTH_LONG).show();
    }

    @Override
    public void callbackWrongCode() {
        showViewsOnScreen();
        showWrongCode();
        codeInput.setError("Неправильный код");
        codeInput.requestFocus();
    }

    private void goBackToAuthorization() {
        super.onBackPressed();
    }

}

File: app/src/main/java/com/bunbeauty/ideal/myapplication/cleanArchitecture/mvp/activities/logIn/VerifyPhoneActivity.kt
package com.bunbeauty.ideal.myapplication.cleanArchitecture.mvp.activities.logIn

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import com.android.ideal.myapplication.R
import com.arellomobile.mvp.MvpAppCompatActivity
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.bunbeauty.ideal.myapplication.cleanArchitecture.business.logIn.VerifyPhoneInteractor
import com.bunbeauty.ideal.myapplication.cleanArchitecture.di.AppModule
import com.bunbeauty.ideal.myapplication.cleanArchitecture.di.DaggerAppComponent
import com.bunbeauty.ideal.myapplication.cleanArchitecture.mvp.activities.profile.ProfileActivity
import com.bunbeauty.ideal.myapplication.cleanArchitecture.mvp.presenters.VerifyPhonePresenter
import com.bunbeauty.ideal.myapplication.cleanArchitecture.mvp.views.VerifyPhoneView
import com.bunbeauty.ideal.myapplication.helpApi.WorkWithViewApi
import com.google.firebase.auth.PhoneAuthProvider
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import kotlin.reflect.typeOf

class VerifyPhoneActivity : MvpAppCompatActivity(), View.OnClickListener, VerifyPhoneView {

    private val USER_PHONE = "user phone"

    private lateinit var verifyCodeBtn: Button
    private lateinit var resendCodeText: TextView
    private lateinit var codeInput: EditText
    private lateinit var changePhoneText: TextView
    private lateinit var alertCodeText: TextView
    private lateinit var progressBar: ProgressBar

    @Inject
    internal lateinit var verifyPhoneInteractor: VerifyPhoneInteractor

    @InjectPresenter
    internal lateinit var verifyPhonePresenter: VerifyPhonePresenter

    @ProvidePresenter
    internal fun provideVerifyPhonePresenter(): VerifyPhonePresenter {

        DaggerAppComponent.builder()
                .appModule(AppModule(application, intent))
                .build()
                .inject(this)
        return VerifyPhonePresenter(verifyPhoneInteractor)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.verify_phone)
        initView()
        showViewsOnScreen()
        verifyPhonePresenter.sendCode()
    }

    private fun initView() {
        Log.d(TAG, "initView VerifyPhoneActivity: ")

        verifyCodeBtn = findViewById(R.id.verifyVerifyPhoneBtn)
        resendCodeText = findViewById(R.id.resendCodeVerifyPhoneText)
        alertCodeText = findViewById(R.id.alertCodeVerifyText)
        codeInput = findViewById(R.id.codeVerifyPhoneInput)
        changePhoneText = findViewById(R.id.changePhoneVerifyPhoneText)

        progressBar = findViewById(R.id.loadingVerifyPhoneProgressBar)

        verifyCodeBtn.setOnClickListener(this)
        resendCodeText.setOnClickListener(this)
        changePhoneText.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        WorkWithViewApi.hideKeyboard(this)
        when (v.id) {
            R.id.verifyVerifyPhoneBtn -> verifyPhonePresenter.verify(codeInput.text.toString())
            R.id.resendCodeVerifyPhoneText -> verifyPhonePresenter.resendCode(this)
            R.id.changePhoneVerifyPhoneText -> goBackToAuthorization()
        }
    }

    override fun hideViewsOnScreen() {
        verifyCodeBtn.visibility = View.GONE
        resendCodeText.visibility = View.GONE
        codeInput.visibility = View.GONE
        changePhoneText.visibility = View.GONE
        alertCodeText.visibility = View.GONE
        progressBar.visibility = View.VISIBLE
    }

    override fun showViewsOnScreen() {
        verifyCodeBtn.visibility = View.VISIBLE
        resendCodeText.visibility = View.VISIBLE
        codeInput.visibility = View.VISIBLE
        changePhoneText.visibility = View.VISIBLE
        alertCodeText.visibility = View.VISIBLE
        progressBar.visibility = View.GONE
    }

    override fun showSendCode() {
        Toast.makeText(this, "Код был отправлен", Toast.LENGTH_LONG).show()
    }

    override fun showWrongCode() {
        Toast.makeText(this, "Вы ввели неверный код", Toast.LENGTH_LONG).show()
    }

    override fun callbackWrongCode() {
        showViewsOnScreen()
        showWrongCode()
        codeInput.error = "Неправильный код"
        codeInput.requestFocus()
    }

    private fun goBackToAuthorization() {
        super.onBackPressed()
    }

    override fun goToRegistration(phone: String) {
        val intent = Intent(this, RegistrationActivity::class.java)
        intent.putExtra(USER_PHONE, phone)
        startActivity(intent)
        overridePendingTransition(0,0)
        finish()
    }

    override fun goToProfile() {
        val intent = Intent(this, ProfileActivity::class.java)
        startActivity(intent)
        overridePendingTransition(0,0)
        finish()
    }

    override fun sendVerificationCode(phoneNumber: String, callback: PhoneAuthProvider.OnVerificationStateChangedCallbacks) {
        PhoneAuthProvider.getInstance().verifyPhoneNumber(
                phoneNumber, // Phone number to verify
                60, // Timeout duration
                TimeUnit.SECONDS, // Unit of timeout
                this, // Activity (for callback binding)
                callback)
    }

    companion object {
        private val TAG = "DBInf"
    }
}